### PR TITLE
New, improved, rewritten assembler

### LINKED ISSUE #97: Add back the local assembler
#96 temporarily removes the local assembler. The local assembler should be added back, preferably without a strong reliance on the HMM aligner, which is slow.

### PR DESCRIPTION
**Pull request by fnothaft**
Resolves #97. Eliminates most of the old assembler code. Adds a new assembler that emits observations directly from the de Brujin graph. In the old assembler, we generated haplotypes and realigned the reads to the haplotypes before calculating genotype likelihoods. In this assembler, we can handle longer haplotypes, and we  eliminate the need to realign reads.

### CODE REVIEW

**Comment by pgrosu** [comment]
Hi Frank,

Happy New Year!  Thank you for the nice updates, though  I seem to be getting some weird errors when I run `mvn package`:

```
$ mvn package
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO]
[INFO] avocado: A Variant Caller, Distributed
[INFO] avocado: sufficient statistics packaging code/format
[INFO] avocado-core: A Variant Caller, Distributed
[INFO]
[INFO] Using the builder org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder with a thread count of 1
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building avocado: A Variant Caller, Distributed 0.0.2
[INFO] ------------------------------------------------------------------------
Downloading: http://hadoop-bam.sourceforge.net/maven/edu/berkeley/cs/amplab/adam/adam-format/0.7.1-SNAPSHOT/maven-metadata.xml
Downloading: http://oss.sonatype.org/content/repositories/snapshots/edu/berkeley/cs/amplab/adam/adam-format/0.7.1-SNAPSHOT/maven-metadata.xml
Downloading: http://people.apache.org/repo/m2-snapshot-repository/edu/berkeley/cs/amplab/adam/adam-format/0.7.1-SNAPSHOT/maven-metadata.xml
Downloading: http://oss.sonatype.org/content/repositories/snapshots/edu/berkeley/cs/amplab/adam/adam-format/0.7.1-SNAPSHOT/adam-format-0.7.1-SNAPSHOT.pom
Downloading: http://hadoop-bam.sourceforge.net/maven/edu/berkeley/cs/amplab/adam/adam-format/0.7.1-SNAPSHOT/adam-format-0.7.1-SNAPSHOT.pom
Downloading: http://people.apache.org/repo/m2-snapshot-repository/edu/berkeley/cs/amplab/adam/adam-format/0.7.1-SNAPSHOT/adam-format-0.7.1-SNAPSHOT.pom
[WARNING] The POM for edu.berkeley.cs.amplab.adam:adam-format:jar:0.7.1-SNAPSHOT is missing, no dependency information available
Downloading: http://oss.sonatype.org/content/repositories/snapshots/edu/berkeley/cs/amplab/adam/adam-core/0.7.1-SNAPSHOT/maven-metadata.xml
Downloading: http://hadoop-bam.sourceforge.net/maven/edu/berkeley/cs/amplab/adam/adam-core/0.7.1-SNAPSHOT/maven-metadata.xml
Downloading: http://people.apache.org/repo/m2-snapshot-repository/edu/berkeley/cs/amplab/adam/adam-core/0.7.1-SNAPSHOT/maven-metadata.xml
Downloading: http://oss.sonatype.org/content/repositories/snapshots/edu/berkeley/cs/amplab/adam/adam-core/0.7.1-SNAPSHOT/adam-core-0.7.1-SNAPSHOT.pom
Downloading: http://hadoop-bam.sourceforge.net/maven/edu/berkeley/cs/amplab/adam/adam-core/0.7.1-SNAPSHOT/adam-core-0.7.1-SNAPSHOT.pom
Downloading: http://people.apache.org/repo/m2-snapshot-repository/edu/berkeley/cs/amplab/adam/adam-core/0.7.1-SNAPSHOT/adam-core-0.7.1-SNAPSHOT.pom
[WARNING] The POM for edu.berkeley.cs.amplab.adam:adam-core:jar:0.7.1-SNAPSHOT is missing, no dependency information available
Downloading: http://oss.sonatype.org/content/repositories/snapshots/edu/berkeley/cs/amplab/adam/adam-cli/0.7.1-SNAPSHOT/maven-metadata.xml
Downloading: http://people.apache.org/repo/m2-snapshot-repository/edu/berkeley/cs/amplab/adam/adam-cli/0.7.1-SNAPSHOT/maven-metadata.xml
Downloading: http://hadoop-bam.sourceforge.net/maven/edu/berkeley/cs/amplab/adam/adam-cli/0.7.1-SNAPSHOT/maven-metadata.xml
Downloading: http://oss.sonatype.org/content/repositories/snapshots/edu/berkeley/cs/amplab/adam/adam-cli/0.7.1-SNAPSHOT/adam-cli-0.7.1-SNAPSHOT.pom
Downloading: http://hadoop-bam.sourceforge.net/maven/edu/berkeley/cs/amplab/adam/adam-cli/0.7.1-SNAPSHOT/adam-cli-0.7.1-SNAPSHOT.pom
Downloading: http://people.apache.org/repo/m2-snapshot-repository/edu/berkeley/cs/amplab/adam/adam-cli/0.7.1-SNAPSHOT/adam-cli-0.7.1-SNAPSHOT.pom
[WARNING] The POM for edu.berkeley.cs.amplab.adam:adam-cli:jar:0.7.1-SNAPSHOT is missing, no dependency information available
Downloading: http://oss.sonatype.org/content/repositories/snapshots/args4j/args4j/2.0.23/args4j-2.0.23.pom
Downloading: http://hadoop-bam.sourceforge.net/maven/args4j/args4j/2.0.23/args4j-2.0.23.pom
Downloading: http://people.apache.org/repo/m2-snapshot-repository/args4j/args4j/2.0.23/args4j-2.0.23.pom
Downloading: http://repo.maven.apache.org/maven2/args4j/args4j/2.0.23/args4j-2.0.23.pom
Downloaded: http://repo.maven.apache.org/maven2/args4j/args4j/2.0.23/args4j-2.0.23.pom (2 KB at 14.7 KB/sec)
Downloading: http://oss.sonatype.org/content/repositories/snapshots/args4j/args4j-site/2.0.23/args4j-site-2.0.23.pom
Downloading: http://hadoop-bam.sourceforge.net/maven/args4j/args4j-site/2.0.23/args4j-site-2.0.23.pom
Downloading: http://people.apache.org/repo/m2-snapshot-repository/args4j/args4j-site/2.0.23/args4j-site-2.0.23.pom
Downloading: http://repo.maven.apache.org/maven2/args4j/args4j-site/2.0.23/args4j-site-2.0.23.pom
Downloaded: http://repo.maven.apache.org/maven2/args4j/args4j-site/2.0.23/args4j-site-2.0.23.pom (5 KB at 149.3 KB/sec)
Downloading: http://oss.sonatype.org/content/repositories/snapshots/org/kohsuke/pom/3/pom-3.pom
Downloading: http://hadoop-bam.sourceforge.net/maven/org/kohsuke/pom/3/pom-3.pom
Downloading: http://people.apache.org/repo/m2-snapshot-repository/org/kohsuke/pom/3/pom-3.pom
Downloading: http://repo.maven.apache.org/maven2/org/kohsuke/pom/3/pom-3.pom
Downloaded: http://repo.maven.apache.org/maven2/org/kohsuke/pom/3/pom-3.pom (4 KB at 124.9 KB/sec)
Downloading: http://oss.sonatype.org/content/repositories/snapshots/edu/berkeley/cs/amplab/adam/adam-format/0.7.1-SNAPSHOT/adam-format-0.7.1-SNAPSHOT.jar
Downloading: http://oss.sonatype.org/content/repositories/snapshots/edu/berkeley/cs/amplab/adam/adam-core/0.7.1-SNAPSHOT/adam-core-0.7.1-SNAPSHOT.jar
Downloading: http://oss.sonatype.org/content/repositories/snapshots/edu/berkeley/cs/amplab/adam/adam-cli/0.7.1-SNAPSHOT/adam-cli-0.7.1-SNAPSHOT.jar
Downloading: http://oss.sonatype.org/content/repositories/snapshots/args4j/args4j/2.0.23/args4j-2.0.23.jar
Downloading: http://hadoop-bam.sourceforge.net/maven/edu/berkeley/cs/amplab/adam/adam-format/0.7.1-SNAPSHOT/adam-format-0.7.1-SNAPSHOT.jar
Downloading: http://hadoop-bam.sourceforge.net/maven/edu/berkeley/cs/amplab/adam/adam-core/0.7.1-SNAPSHOT/adam-core-0.7.1-SNAPSHOT.jar
Downloading: http://hadoop-bam.sourceforge.net/maven/edu/berkeley/cs/amplab/adam/adam-cli/0.7.1-SNAPSHOT/adam-cli-0.7.1-SNAPSHOT.jar
Downloading: http://hadoop-bam.sourceforge.net/maven/args4j/args4j/2.0.23/args4j-2.0.23.jar
Downloading: http://people.apache.org/repo/m2-snapshot-repository/edu/berkeley/cs/amplab/adam/adam-format/0.7.1-SNAPSHOT/adam-format-0.7.1-SNAPSHOT.jar
Downloading: http://people.apache.org/repo/m2-snapshot-repository/edu/berkeley/cs/amplab/adam/adam-core/0.7.1-SNAPSHOT/adam-core-0.7.1-SNAPSHOT.jar
Downloading: http://people.apache.org/repo/m2-snapshot-repository/edu/berkeley/cs/amplab/adam/adam-cli/0.7.1-SNAPSHOT/adam-cli-0.7.1-SNAPSHOT.jar
Downloading: http://people.apache.org/repo/m2-snapshot-repository/args4j/args4j/2.0.23/args4j-2.0.23.jar
Downloading: http://repo.maven.apache.org/maven2/args4j/args4j/2.0.23/args4j-2.0.23.jar
Downloaded: http://repo.maven.apache.org/maven2/args4j/args4j/2.0.23/args4j-2.0.23.jar (66 KB at 1251.3 KB/sec)
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] avocado: A Variant Caller, Distributed ............ FAILURE [ 12.523 s]
[INFO] avocado: sufficient statistics packaging code/format  SKIPPED
[INFO] avocado-core: A Variant Caller, Distributed ....... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 12.768 s
[INFO] Finished at: 2015-01-02T23:25:00-05:00
[INFO] Final Memory: 8M/1042M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project avocado: Could not resolve dependencies for project edu.berkeley.cs.amplab.avocado:avocado:pom:0.0.2: The following artifacts could not be resolved: edu.berkeley.cs.amplab.adam:adam-format:jar:0.7.1-SNAPSHOT, edu.berkeley.cs.amplab.adam:adam-core:jar:0.7.1-SNAPSHOT, edu.berkeley.cs.amplab.adam:adam-cli:jar:0.7.1-SNAPSHOT: Could not find artifact edu.berkeley.cs.amplab.adam:adam-format:jar:0.7.1-SNAPSHOT in Sonatype (http://oss.sonatype.org/content/repositories/snapshots/) -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
$
```
Thanks,
~p

**Comment by fnothaft** [comment]
@pgrosu thanks; happy new years to you too!

As for the compile issue, it looks like you're building from an out-of-date repository. The `edu.berkeley.cs.amplab` dependencies moved to `org.bdgenomics` a bit back; if you pull and update your branch you should be able to compile fine.

**Comment by pgrosu** [comment]
I think I did a little too much celebrating during the New Year, and only ran the clone without the pull.  Running the following steps made it all better :)

```
$ git clone https://github.com/fnothaft/avocado threaded-assembler
$ cd threaded-assembler/
$ git pull https://github.com/fnothaft/avocado threaded-assembler
$ mvn package
```

Thanks Frank!
~p

**Comment by fnothaft** [comment]
Ah! No worries; glad to hear things are working now.

**Comment by fnothaft** [comment]
I added documentation, so this now depends on #129.

File: avocado-core/pom.xml [hunk 1]
```diff
@@ -125,5 +125,21 @@
       <artifactId>scalatest_${scala.version.prefix}</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.seqdoop</groupId>
+      <artifactId>hadoop-bam</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.seqdoop</groupId>
+      <artifactId>cofoja</artifactId>
```
Comment (tdanford) [inline avocado-core/pom.xml]: Is this a separate dist of the cofoja library than what we were depending on elsewhere?  Did cofoja (do we know) ever get itself in working order for Java 1.8?  That was an issue, a while ago...

Comment (fnothaft) [inline avocado-core/pom.xml]: Nah, this is the same Cofoja. Cofoja is a won't fix for Java 8, IIRC, but was removed from the latest version of HTSJDK/Picard. _Howaaaaver,_ no one has cut a new release of HTSJDK yet. See https://github.com/samtools/htsjdk/pull/115#issuecomment-72462083.

File: avocado-core/src/main/scala/org/bdgenomics/avocado/algorithms/debrujin/Kmer.scala [hunk 2]
```diff
@@ -39,6 +49,15 @@ case class Kmer(val kmerSeq: String, val weight: Int = 1, val inFlank: Boolean =
     prefix + "[" + suffix + "]"
   }
 
+  def toDetailedString: String = {
+    kmerSeq + " * " + multiplicity + ", " + refPos.fold("unmapped")("@ " + _) + "\n" +
```
Comment (tdanford) [inline avocado-core/src/main/scala/org/bdgenomics/avocado/algorithms/debrujin/Kmer.scala]: Can't you use ```mkString``` instead of these folds?

Comment (fnothaft) [inline avocado-core/src/main/scala/org/bdgenomics/avocado/algorithms/debrujin/Kmer.scala]: Good point; `mkString` would be preferable with the exception of the `refPos.fold`.

**Comment by fnothaft** [comment]
Jenkins, retest this please.

File: avocado-core/src/main/scala/org/bdgenomics/avocado/algorithms/debrujin/Kmer.scala (hunk 2 above)
Comment (fnothaft) [inline avocado-core/src/main/scala/org/bdgenomics/avocado/algorithms/debrujin/Kmer.scala]: Updated.

File: avocado-core/src/main/scala/org/bdgenomics/avocado/algorithms/debrujin/KmerGraph.scala [hunk 3]
```diff
@@ -119,213 +41,561 @@ object KmerGraph {
...
+  // use default alignment parameters, except infinite padding penalty
+  private val hmm = new HMMAligner(new TransitionMatrixConfiguration(LOG_PADDING_PENALTY = Double.NegativeInfinity))
 
-  def addFlanks(reference: String, flankLength: Int) {
-    val startFlank = reference.take(flankLength)
-    val endFlank = reference.takeRight(flankLength)
-    val kmerSequences = KmerGraph.getKmerSequences(startFlank, kmerLength) ++ KmerGraph.getKmerSequences(endFlank, kmerLength)
-    addSequencesToGraph(kmerSequences, true)
+  override def toString(): String = {
+    "Sources: " + sourceKmers.map(_.kmerSeq).reduce(_ + ", " + _) + "\n" +
```
Comment (tdanford) [inline avocado-core/src/main/scala/org/bdgenomics/avocado/algorithms/debrujin/KmerGraph.scala]: same mkString comment as above.

**Comment by fnothaft** [comment]
Rebased on #131 and added an active region threshold.

**Comment by fnothaft** [comment]
Jenkins, retest this please.

**Pull request closed without merging**